### DISCUSSION TITLE
Time Series | Creating Custom Models & Pipelines

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1867] [FAQs]
# Suggested Previous Reads

   - **[PyCaret Time Series Module Architecture Overview](https://medium.com/towards-data-science/pycaret-time-series-module-architecture-overview-57336a2f39c7)**

# Resources for this FAQ

- **[Jupyter Notebook](https://nbviewer.org/github/ngupta23/pycaret_faqs/blob/main/time_series/pycaret_ts_custom_model_sktime.ipynb)** (can be opened in Google Colab)

# Overview

The `pycaret` timeseries module currently supports around 30 models/pipelines natively. However, there may be cases where the users may want to use their own pipeline to do time series analysis. Well, now we can! `pycaret` allows users to add their own custom `sktime` models and pipelines into the `pycaret` framework. Let's see how.

# Solution

For the purpose of this illustration, we will use the famous airline dataset and simulate some missing values. Then, we will try to model this with a custom `sktime` pipeline.

```python

#### Get Data ----
y = get_data("airline", verbose=False)

#### Simulate missing values ----
y[60:72] = np.nan
```

Next, let's create a sktime pipeline with preprocessing. This includes an imputer and a log transformer. The log transformer is not nexessary, but just added to show the capability. There are many other transformers available as well. **Check out https://www.sktime.org/en/stable/api_reference/transformations.html** for details about all the transformers that are available.

```python
#### Define your custom pipeline ----

from sktime.transformations.series.impute import Imputer
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.forecasting.arima import ARIMA

my_forecaster = TransformedTargetForecaster(
    [
        ("impute", Imputer()),
        ("log", LogTransformer()),
        ("model", ARIMA(seasonal_order=(0,1,0,12))),
    ]
)
```

Next, let's model this custom pipeline with `pycaret`. Modeling is done via the same `create_model` call that is used for native `pycaret` models.

```python
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, session_id=42)

my_custom_model = exp.create_model(my_forecaster)
```

![image](https://user-images.githubusercontent.com/33585645/142617330-aa511118-aca9-4176-b76d-81f1b421788d.png)

```python
exp.plot_model(my_custom_model)
```

![image](https://user-images.githubusercontent.com/33585645/142618382-7066b8c3-839e-40f9-ac72-8f34a8901eee.png)

```python
my_custom_model
```

![image](https://user-images.githubusercontent.com/33585645/142618440-3c60150e-e230-4ed0-aca3-d0efb697e526.png)

# Conclusion

While pycaret provides inbuilt models/pipelines for time series analysis, users are by no means limited to using just those models/pipelines. The flexibility to add user defined models/pipelines is what makes it `pycaret` really shine. So next time you do not find a model/pipeline in pycaret, try adding your own custom one. 

That's it for now. Happy forecasting!

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1954

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
